mdbook-tera-backend/src/tera_renderer/custom_component.rs

```diff
@@ -0,0 +1,133 @@
+            .as_str()
+            .ok_or_else(|| tera::Error::from(format!("S has invalid type, expected string")))?;
+        let prefix = args
+            .get("prefix")
+            .ok_or_else(|| tera::Error::from(format!("No prefix argument provided")))?
+            .as_str()
+            .ok_or_else(|| {
+                tera::Error::from(format!("Prefix has invalid type, expected string"))
+            })?;
+        string
+            .strip_prefix(prefix)
+            .map(|s| tera::Value::String(s.to_owned()))
+            .ok_or_else(|| tera::Error::from(format!("Could not strip prefix")))
+    }
+}
+
+pub struct CustomComponent {
+    template: Tera,
+    name: String,
+    /// Used to generate unique ids for each component to prevent collisions in javascript with query selectors.
+    counter: RefCell<u32>,
+}
+
+impl CustomComponent {
+    pub fn new(name: &str, template_str: &str, dependencies: &[&Self]) -> Result<CustomComponent> {
+        let mut template = Tera::default();
+        for dep in dependencies {
+            template.extend(&dep.template)?;
+        }
+        template.add_raw_template(name, template_str)?;
+        template.register_function("strip_prefix", make_strip_prefix_function());
+
+        Ok(CustomComponent {
+            name: String::from(name),
+            counter: RefCell::new(0),
+            template,
+        })
+    }
+
+    pub fn register_function(&mut self, name: &str, function: impl tera::Function + 'static) {
+        self.template.register_function(name, function);
+    }
+
+    fn create_context(
+        &self,
+        rendering_context: &RenderingContext,
+        attributes: BTreeMap<String, String>,
+    ) -> tera::Context {
+        let counter = self.counter.replace_with(|&mut counter| counter + 1);
+        let mut context = tera::Context::new();
+        context.insert("counter", &counter);
+        context.insert("language", &rendering_context.language);
+        context.insert("path", &rendering_context.path);
+        context.insert("ctx", &rendering_context.serialized_ctx);
+        context.insert(
+            "book_dir",
+            &rendering_context.ctx.destination.parent().unwrap(),
+        );
+        context.insert("attributes", &attributes);
+
+        context
+    }
+
+    pub fn render(
+        &self,
+        rendering_context: &RenderingContext,
+        attributes: BTreeMap<String, String>,
+    ) -> Result<String> {
+        let context = self.create_context(rendering_context, attributes);
+        let output = self.template.render(&self.name, &context)?;
+        Ok(output)
+    }
+
+    pub fn component_name(&self) -> String {
+        self.name.clone()
+    }
+}
+
+#[derive(Deserialize)]
+pub struct TeraComponentConfig {
+    pub name: String,
+    pub path: PathBuf,
+
+    #[serde(default)]
+    pub dependencies: Vec<String>,
+}
+
+#[derive(Deserialize)]
+pub struct TeraRendererConfig {
+    pub components: Vec<TeraComponentConfig>,
+}
```

Comment:
This is the kind of infrastructure that I don't expect there. In my mind, the renderer should just iterate over the HTML files and process the one by one. It will probably be a 20-30 line renderer at the end :slightly_smiling_face:

Reply:
It's needed to specify the path to templates to include and their dependencies, it is required by Tera

Reply:
Are you saying that a Tera template cannot include files which haven't been explicitly configured in Tera somehow? Can we use an include directory instead? Just define that all files in `theme/templates/` are available for use in the templates. That should simplify the setup.

Reply:
It works too, the tradeoff is that we'll need to read the directory and extract file names from there, and since all templates will have access to one another, they could have cyclical dependencies (which is handled by tera)